### PR TITLE
feat: add Safaricom M-PESA (Daraja) payment provider plugin

### LINKED ISSUE #696: [Feature Request]: Mobile Money Provider Integration - M-Pesa Kenya (SafariCom/Vodafone)
### Preflight Checklist

- [x] I have read the [Contributing Guidelines](https://github.com/EnAccess/micropowermanager/blob/main/CONTRIBUTING.md) for this project, if it exists.
- [x] I agree to follow the [Code of Conduct](https://github.com/EnAccess/micropowermanager/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [x] I have searched the [issue tracker](https://github.com/EnAccess/micropowermanager/issues) for a feature request that matches the one I want to file, without success.

### Problem Description

TBD

### Proposed Solution

TBD

### Alternatives Considered

TBD

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by AbogeJr**
Adds a payment-provider plugin for Safaricom's Daraja v3 API so MPM deployments in Kenya can accept M-PESA payments via STK Push (M-Pesa Express)

### Brief summary of the change made
Adds a Safaricom M-PESA payment-provider plugin that lets mini-grid operators in Kenya accept payments via STK Push (M-Pesa Express / Daraja v3). Unlike the existing hosted-redirect providers (Paystack, PesaPal), the customer never leaves MPM, the operator initiates from Safaricom M-PESA → Initiate Payment, the customer's phone receives the M-PESA prompt, they enter their PIN, and the page polls Daraja's STK Push Query endpoint until the transaction resolves.

closes: #696 

### Describe how you tested your changes?

Manual end-to-end against Daraja sandbox:

- Created a new Daraja app under the M-Pesa Express product on developer.safaricom.co.ke and copied the Consumer Key/Secret.
- Enabled the plugin in MPM, opened Safaricom M-PESA → Credentials, pasted the key/secret, left Shortcode and Passkey blank (sandbox fallback), set Environment to Sandbox, saved.
- Overview page confirmed status flipped to "Configured (Sandbox)" in green.
- Initiate Payment page: picked Device Type = Meter, typed a known meter serial, saw the green ✓ "Meter found and in use." validation hint. Phone 0712737135, amount 1. Submitted.
- Confirmed the M-PESA STK Push prompt arrived on the test phone within 3s.
- Watched the Waiting state poll Daraja every 3s with the countdown updating; on PIN entry, the page flipped to "Payment Successful" with the M-PESA receipt number.
- Negative paths verified: cancelled the prompt (got the "Cancelled by Customer" result with Daraja code 1032, transaction recorded as STATUS_ABANDONED);

Screenshot from 2026-06-19 16-57-11
Screenshot from 2026-06-19 16-13-07
Screenshot from 2026-06-19 16-12-58
Screenshot from 2026-06-19 16-12-24
Screenshot from 2026-06-19 16-12-08
Screenshot from 2026-06-19 16-10-49
Screenshot_20260619-161215_Pixel Launcher

### Pull Request checklist
Please confirm you have completed any of the necessary steps below.

- [x] Meaningful Pull Request title and description
- [x] Changes tested as described above
- [x] Added appropriate documentation for the change.
- [x] Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: docs/integrations-guide/safaricom-mobile-money-integration.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+> [!INFO]
+> The **operator** initiates STK Push from the MPM sidebar (**Safaricom M-PESA → Initiate Payment**). There is no customer-facing public payment URL for M-PESA.
+
+## Overview
+
+### Pre-requisites
+
+1. Access to the MPM admin panel.
+2. A [Safaricom Developer account](https://developer.safaricom.co.ke/) (free).
+3. A Daraja **app** with the **M-Pesa Express** (Lipa Na M-Pesa Online) product enabled, and the **Consumer Key** and **Consumer Secret** that app issues.
```
Comment (dmohns) [inline docs/integrations-guide/safaricom-mobile-money-integration.md]: Why does is `Lipa Na M-Pesa` mentioned here? Earlier you are saying we are currently only doing MPM-initiated payments. But Lipa Na M-Pesa would be customer initiated payments via till number, no?

File: docs/integrations-guide/safaricom-mobile-money-integration.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+3. Click **Save**.
+4. The Configuration status box on the Overview page should turn green and show **Configured (Sandbox)**.
+
+> [!INFO]
+> In sandbox the plugin pairs your Consumer Key/Secret with Daraja's public sandbox shortcode `174379` and the published LNM passkey `bfb279f9aa9bdbcf158e97dd71a467cd2e0c893059b10f78e6b72ada1ed2c919`. This combination is the only one Daraja sandbox accepts. Custom shortcodes will fail with `Merchant does not exist`.
+
+### Step 5: Verify the Webhook URL is Reachable
+
+For the asynchronous callback path to work, Daraja must be able to POST to your `CallBackURL` over HTTPS. Local development without a tunnel will not receive callbacks — but the plugin also polls Daraja's STK Push Query endpoint, so transactions still resolve correctly even when Daraja can't reach you.
+
```
Comment (dmohns) [inline docs/integrations-guide/safaricom-mobile-money-integration.md:93]: We already have a general section covering this! Maybe you can link it? https://micropowermanager.io/development/development-environment.html#api-gateway-with-ngrok

File: src/backend/app/Plugins/SafaricomMobileMoney/Models/SafaricomCredential.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        return $this->consumer_secret ?? '';
+    }
+
+    public function getPasskey(): string {
+        return $this->passkey ?? '';
+    }
+
+    public function getShortcode(): string {
+        return $this->shortcode ?? '';
+    }
```
Comment (dmohns) [inline src/backend/app/Plugins/SafaricomMobileMoney/Models/SafaricomCredential.php]: We recently started using PHP's asym visibility over trivial getters, see https://github.com/EnAccess/micropowermanager/pull/1516

If you need (for example `passkey`) to be read only, can you use

```php
public protected(set) string $passkey;
```

instead?

File: src/backend/app/Plugins/SafaricomMobileMoney/Models/SafaricomTransaction.php [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+    protected $casts = [
+        'metadata' => 'array',
+        'response_data' => 'array',
+        'transaction_date' => 'datetime',
+    ];
+
+    public function getAmount(): float {
+        return $this->amount;
+    }
```
Comment (dmohns) [inline src/backend/app/Plugins/SafaricomMobileMoney/Models/SafaricomTransaction.php]: As above, let's try to reduce use of trivial getters and setters.

File: src/backend/app/Plugins/SafaricomKePaymentProvider/Services/SafaricomTransactionService.php [hunk 5]
```diff
@@ -0,0 +1,677 @@
...
+        }
+        if (preg_match('/^0\d{9}$/', $digits) === 1) {
+            return '254'.substr($digits, 1);
+        }
+        if (preg_match('/^[17]\d{8}$/', $digits) === 1) {
+            return '254'.$digits;
+        }
+
+        return null;
+    }
```
Comment (dmohns) [inline src/backend/app/Plugins/SafaricomKePaymentProvider/Services/SafaricomTransactionService.php:590]: Can you utilize https://github.com/giggsey/libphonenumber-for-php-lite here? Which we also use in other places to handle phone numbers instead of chaining regex?

File: src/backend/app/Plugins/SafaricomMobileMoney/Models/SafaricomTransaction.php [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+    /**
+     * @return MorphMany<TransactionConflicts, $this>
+     */
+    public function conflicts(): MorphMany {
+        return $this->morphMany(TransactionConflicts::class, 'transaction');
+    }
+
+    public static function getTransactionName(): string {
+        return self::RELATION_NAME;
+    }
```
Comment (dmohns) [inline src/backend/app/Plugins/SafaricomMobileMoney/Models/SafaricomTransaction.php]: I don't think you need those anymore, they are already available via the base class https://github.com/EnAccess/micropowermanager/pull/1511

**Review by dmohns — COMMENTED** [review]
Hey @AbogeJr 

amazing PR! I'm leaving a lot of comments and suggestions here 🙏 

- Could you rename the plugin to `Safaricom Kenya` (or KE, what every you prefer) to indicate that the integration is specific to Kenya.

- Could you clarify a bit more about the async webhook part? It looks like using a webhook is entirely optional, as your code will use status polling anyways. Also, you mention the webhook has proven to be unreliable in your testing? What is the point of using it at all then?

File: docs/integrations-guide/safaricom-mobile-money-integration.md (hunk 1 above)
Comment (AbogeJr) [reply docs/integrations-guide/safaricom-mobile-money-integration.md]: True. Most of the guides and articles (especially the older ones) use the terms interchangeably but I now see that could get confusing in this case. I'll stick with M-Pesa Express since that is the official service name on Daraja 3.0

**Comment by AbogeJr** [comment]
> Hey @AbogeJr
> 
> amazing PR! I'm leaving a lot of comments and suggestions here 🙏
> 
> * Could you rename the plugin to `Safaricom Kenya` (or KE, what every you prefer) to indicate that the integration is specific to Kenya.
> * Could you clarify a bit more about the async webhook part? It looks like using a webhook is entirely optional, as your code will use status polling anyways. Also, you mention the webhook has proven to be unreliable in your testing? What is the point of using it at all then?

About the webhook, it is the recommended way to get the transaction status. In my initial tests, it was taking longer than expected in some runs. However, with polling alone there is an edge case where if the operator closes the transaction page while waiting for status update, it will stay in STATUS_REQUESTED state forever. 

From my current tests (with tunneling), the polling ensures the status is updated immediately, and if the operator closes/refreshes, the transaction status is still updated asynchronoutsly.

File: src/backend/database/migrations/2025_06_08_210412_add_safaricom_ke_payment_provider_to_mpm_plugin_table.php [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+use Illuminate\Support\Facades\DB;
+
+return new class extends Migration {
+    public function up(): void {
+        DB::table('mpm_plugins')->insert([
+            [
+                'id' => MpmPlugin::SAFARICOM_KE_PAYMENT_PROVIDER,
+                'name' => 'SafaricomKePaymentProvider',
+                'description' => 'Safaricom M-PESA integration for MicroPowerManager.',
+                'tail_tag' => 'Safaricom M-PESA',
```
Comment (dmohns) [inline src/backend/database/migrations/2025_06_08_210412_add_safaricom_ke_payment_provider_to_mpm_plugin_table.php]: There is a semantical merge conflict here with https://github.com/EnAccess/micropowermanager/pull/1526 (where `tail_tag` was removed).

Currently this migration fails with 

```
backend-dev       |   SQLSTATE[42S22]: Column not found: 1054 Unknown column 'tail_tag' in 'field list' (Connection: micro_power_manager, Host: db, Port: 3306, Database: micro_power_manager, SQL: insert into `mpm_plugins` (`created_at`, `description`, `id`, `installation_command`, `name`, `root_class`, `tail_tag`, `updated_at`) values (2026-06-29 14:04:31, Safaricom M-PESA integration for MicroPowerManager., 31, safaricom-ke-payment-provider:install, SafaricomKePaymentProvider, SafaricomKePaymentProvider, Safaricom M-PESA, 2026-06-29 14:04:31))
```

**Comment by dmohns** [comment]
image

When a transaction gets rejected this record is persisted.

- Can you make sure to always persist phone numbers in E164 format? Ideally you use libphonenumber 
- The icon for Safaricom is missing

File: docs/integrations-guide/safaricom-mobile-money-integration.md [hunk 8]
```diff
@@ -82,15 +82,11 @@ Unlike hosted-checkout providers (Paystack, PesaPal), M-PESA uses **STK Push** (
 > [!INFO]
 > In sandbox the plugin pairs your Consumer Key/Secret with Daraja's public sandbox shortcode `174379` and the published LNM passkey `bfb279f9aa9bdbcf158e97dd71a467cd2e0c893059b10f78e6b72ada1ed2c919`. This combination is the only one Daraja sandbox accepts. Custom shortcodes will fail with `Merchant does not exist`.
 
-### Step 5: Verify the Webhook URL is Reachable
+### Step 5: (Optional) Expose Your Webhook to Daraja
 
-For the asynchronous callback path to work, Daraja must be able to POST to your `CallBackURL` over HTTPS. Local development without a tunnel will not receive callbacks — but the plugin also polls Daraja's STK Push Query endpoint, so transactions still resolve correctly even when Daraja can't reach you.
+**You can skip this step.** The plugin polls Daraja's STK Push Query endpoint while a transaction is pending, so every status — success, insufficient funds, cancelled, timeout — resolves correctly even when Daraja can't reach your backend. The webhook is a faster path to the same result, not a required one.
```
Comment (dmohns) [inline docs/integrations-guide/safaricom-mobile-money-integration.md]: This step seem absolutely mandatory! In my testing if I skip this step it will send `localhost` as callback URL which leads to 

```
backend-dev       | [2026-06-29 14:33:41] development.ERROR: Safaricom STK Push rejected by Daraja {"reference_id":"33650dba-8501-49ec-98c4-97e7fadd6c7f","shortcode":"174379","status":400,"body_excerpt":"{
backend-dev       |                     \"requestId\":\"0bea-4645-bbd5-d45e2b4abeaf23630\",
backend-dev       |                     \"errorCode\": \"400.002.02\",
backend-dev       |                     \"errorMessage\": \"Bad Request - Invalid CallBackURL\"
backend-dev       |                 }
backend-dev       |             "} 
```

File: docs/installation/configuration-production.md [hunk 9]
```diff
@@ -106,9 +106,9 @@ MicroPowerManager uses Laravel's Trusted Proxy feature to correctly handle reque
 
    ```yaml
    # In your ConfigMap (e.g. k8s/base/gcp_gke/configmaps.yaml)
-   TRUSTEDPROXY_PROXIES: 35.191.0.0/16,130.211.0.0/22  # GCP load balancer IP ranges
+   TRUSTEDPROXY_PROXIES: 35.191.0.0/16,130.211.0.0/22 # GCP load balancer IP ranges
    # For AWS, use the appropriate AWS ELB IP ranges or '*', if you understand the risks
-   TRUSTEDPROXY_PROXIES: '*'  # Trust all proxies (not recommended for production)
+   TRUSTEDPROXY_PROXIES: "*" # Trust all proxies (not recommended for production)
    ```
```
Comment (dmohns) [inline docs/installation/configuration-production.md:112]: Could you merge the latest `main`. It's been resolved here: https://github.com/EnAccess/micropowermanager/pull/1538

**Comment by dmohns** [comment]
In the frontend you are introducing a lot of custom CSS and styles. And to make it work the use of `::v-deep` + `!important` which produces generally very hard to main code.

```js
::v-deep .safaricom-status-chip.md-error {
  background-color: #d9534f !important;
  color: #fff !important;
}
```

@beesaferoot I think we should add to our FE's `Agents.md` to general discourage the use of `::v-deep` and forbid the use of `!important`. WDYT?

As a result the UI is inconsistent to the rest of MPM

**Round corners**

image

**Inconsistent header**

image

While is personally agree that your components look better, I would prefer to add as little customisation as possible to the codebase. Could you please refactor your FE components to re-use as many existing FE elements and design as possible?

**Comment by beesaferoot** [comment]
The differing UI in part is my fault 🙈, Paystack introduced this style of UI which I assume is being borrowed here. 
The PesaPal plugin UI is based of this as well. 

It might be better to apply the refactor of this style disparity in one refactor PR.  What do you think?

File: docs/integrations-guide/safaricom-mobile-money-integration.md (hunk 8 above)
Comment (AbogeJr) [reply docs/integrations-guide/safaricom-mobile-money-integration.md]: Oh I've seen it now. My local env APP_URL was set the whole time even when I was not tunneling so I wan't getting the error. Updating the docs, thanks.

**Comment by AbogeJr** [comment]
image


@dmohns 
Hey. Thanks for the extensive feedback.


I have addressed all the issues except for this cc @beesaferoot 
>The differing UI in part is my fault 🙈, Paystack introduced this style of UI which I assume is being borrowed here.
>The PesaPal plugin UI is based of this as well.
>
>It might be better to apply the refactor of this style disparity in one refactor PR. What do you think?

Yeah I agree, a separate PR to align with the UI on Paystack, Pesapal and Safaricom plugins might be better.

**Comment by dmohns** [comment]
> > The differing UI in part is my fault 🙈, Paystack introduced this style of UI which I assume is being borrowed here.
> > The PesaPal plugin UI is based of this as well.
> > It might be better to apply the refactor of this style disparity in one refactor PR. What do you think?
> 
> Yeah I agree, a separate PR to align with the UI on Paystack, Pesapal and Safaricom plugins might be better.

Alright, perfect! Let's leave the UI like is then.

**Review by dmohns — APPROVED** [review]
Really nice contribution. Thank you very much @AbogeJr

**Comment by AbogeJr** [comment]
> Really nice contribution. Thank you very much @AbogeJr 

Glad I could help!
Thanks🙏 @dmohns @beesaferoot
